### ISSUE TITLE
Devise a strategy and write code to optimize WU assignments

### ISSUE THREAD

**Issue by bb30994** [FoldingAtHome/fah-issues#1479]
> ...Surely the software only sends smaller units to weaker cards and bigger ones to more powerful ones... is that the case?...

Surely that's what SHOULD happen, but it's still on the wish-list/enhancement list except in rather limited situations.

**Comment by shorttack** [comment]
@jcoffland That's what assignment server does: send appropriate units based on gpu power. Right?

**Comment by PantherX** [comment]
@shorttack 
AFAIK, the AS assigned the WS based on the reported species of the GPU. It is done on the GPU architecture, not on it's model/performance.

**Comment by informatorius** [comment]
There is always an overlap of GPU performance of old gen GPU and new gen GPU. So GPU species should get mixed across architectures. Low atom count is currently ~25k. Mid atom count is ~65k. High atom count is ~150k.
1) The fast GPUs of new gen should get the high atom count work units preferred. 
2) The mid GPUs of new gen and the fast GPUs of old gen should get the mid atom count work units preferred.
3) The slow GPUs of new gen and the mid GPUs of old gen should get the low atom count work units preferred.
4) the slow GPUs of old gen should also get the low atom count work units. (And will in future get too slow to complete them in time so will get blacklisted in future.)

**Comment by shorttack** [comment]
> There is always an overlap of GPU performance of old gen GPU and new gen GPU. So GPU species should get mixed across architectures.

I agree mis-allocating work units is a problem and a waste of resources. Our solution needs to be scalable over time. Two variables that are common across architectures and generations are number of 32-bit cores (Cores) and base core clock in MHz (Clock). I'd recalibrate species based on a Core * Clock algorithm.

I would put this overdue enhancement **on the 2020 plan to complete**. It likely will **take new development resources.**

**Comment by shorttack** [comment]
On reflection, this enhancement project could be **handed off with a fork to one or more of our new Folding Partners such as nVidia, AMD, or Intel**. They have backend Dev resources we don't have to spare.

**Comment by PantherX** [comment]
BTW, the one issue is that you assume that the new generation of GPUs will be faster than the previous one. That might be mostly true for gaming but for F@H, it's very different. Previously, there have been cases were the new GPUs were not supported for months. In others, the new GPU would be running inefficiently than the previous gen GPU. It would take a new FahCore version to fix that. In other cases, the performance difference didn't matter for folding between the two generations. While AMD/Nvidia focuses on gaming, computing performance can't be based on gaming performance.

**Comment by gchernis** [comment]
Could this ranking be done based on FAHBench results?  Even better, could this ranking be done based on apparent performance? F@H could aggregate performance data as CPUs and GPUs complete WUs, then feed that back into assignment server performance rating tables.

**Comment by progger-dev** [comment]
I'm really glad you suggested that @gchernis. I was about suggest it myself. I assume that there are overclockers in the community that do a better job of configuring their machines for better performance. I really think that it would be awesome if the assignment could take into account the performance of each individual client into account.

If you need resources, I would be more than happy to help with this project.

**Comment by jchodera** [comment]
Just a heads-up that we're working on extensive benchmarks of a variety of systems to be able to gather data to be able to predict performance on individual GPUs based on some easily-computed features of any project.

**Comment by gchernis** [comment]
@progger-dev As I'm sure you know, matching is a hard problem in computer science.  How would you match a work unit from project pool to worker from a donor pool?  It cannot be done perfectly, but best is the enemy of good.

One such way could be by memory requirements.  That is, matching atom counts to GPU memory size or PC RAM size,  Could assignment server take available memory as baseline only?

Another such way would be by processing capabilities.  Assignment server should start at some baseline for a each donor's GPU and CPU, based on what's known from the model number.  Then as WUs are completed and submitted, assignment server should remember a short sliding window of WUs per project, converge on actual performance over time, and yield better matching.
___________________

Right now, it's hard to know if donor's hardware is performing well.  I have two GPUs, one from NVidia and another from AMD, that have that property.  Thermal throttling out of the box, but running Project 14201 full-out after a little tuning.  Full disclosure: I did not know they are running sub-optimally until I fired up GPU-Z.

To me, thermal and power throttling of modern GPUs and CPUs when they run at full load should be exposed to the user right in the logs and UI.  An additional incentive could come in a form of comparative ranking within the hardware model, such as a reported percentile or a histogram.

This way they'd know to upgrade cooling or power supply to yield faster completion of assigned WUs.  Sometimes, a VBIOS update from TechPowerup! or slightly more aggressive GPU fan settings can result in dramatic increase of WU production, coupled with longer GPU life due to lower temperatures, at a cost of louder fans.

**Comment by gchernis** [comment]
@jchodera What related information flows from FAHClient to Assignment Server?  What additional information could be sent to facilitate WU assignment optimization?

**Comment by bb30994** [comment]
Here's an example from today's report from the owner of a RTX2060 who was assigned a WU with only 22000 atoms.  Note that he's seeing his GPU only 75% utilized.  He'd be much happier if his GPU was 99% busy but there just aren't enough parallelizable calculations to keep 2000 shaders very busy.  If his GPU had fewer shaders, the utilization would be higher even though the WU would actually be completed more slowly.  He'd be much happier if it was 99% utilized.

https://foldingforum.org/viewtopic.php?f=19&t=35663

**Comment by atlbike** [comment]
Hi, I`m one of the whiners.  When I see a low PPD it causes me to envision a cavalier grad student that didn`t benchmark their code properly.  From a donor perspective, it`s critical to fell that your efforts aren`t being squandered or futile.  Last week I created a spreadsheet for the top 50 teams' active retention rate.  Our team (MicroCenter) was #7 at 37%, taking a hard look at that info, about 85% of the top teams have lost more than 2/3 of their donors.  The FAHclient is the front line of retention IMO, many  things that could help,

**Comment by SparklyIsHere** [comment]
If there are going to be put a lot of redesign effort into optimising WU distribution, then my take on it is that it should be made automatic and dynamic on the server side, based on continued gathered statistics from already completed WUs and which hardware it ran on.

This should be made in a way that removes the whole CPU vs GPU vs New vs Old vs whatever discussion, and be made to dynamically assign WU sizes across the board, regardless of hardware, project and atom count, but with the possibility to set priority parameters on both hardware and project assignments.

Basically the WU distribution could be handled by sending bigger or smaller chunks of the same project to the different hardware requesting work, so if a low capability hardware requested work, then the server could choose to only send a WU containing 10 frames to calculate from the project it wants progress on, while a high end hardware could get 1000 frames of work from the same project.

What to send would be based on the gathered statistics and continued benchmarking of all returning WUs, thus enabling you to set a WU size distribution parameter on the server side that says a WU chunk created and sent to the hardware requesting work should on average be able to complete in 3 hours or whatever is most useful.
 
Priority parameters on the server could include Atom count, since you typically do not want to send low atom count projects to higher end GPUs, shader count and type, memory needed, calculation type, minimum frame count, project priority etc. etc., enabling you to send WUs to the generally most suited hardware, but also enabling you to get the projects done, without the need to split projects on CPU or GPU, or redistribute between them, since project progress would be handled automatically, maybe based on a maximum desired completion time parameter and other priority settings, regardless of the ever changing hardware situation in the network.

**Comment by bb30994** [comment]
Another opinion:
https://foldingforum.org/viewtopic.php?p=331724#p331724
Subject: Algorithm for distributing Work Units, statistical model

**Comment by atlbike** [comment]
> Just a heads-up that we're working on extensive benchmarks of a variety of systems to be able to gather data to be able to predict performance on individual GPUs based on some easily-computed features of any project.

I view this as a critical tech to assign correspondingly keyed WUs.  It also opens the door for donors to select (buy) or re-purpose older hardware to run FAH knowing it will be more effectively utilized.  Potentially, it could be used too augment the eSports potential of F@H and points ranking.  I see folding as having a left-brain:right-brain bifurcation of what attracts people to support the cause.

**Comment by PantherX** [comment]
FYI, an update on this is that there's discussion and planning going on to automatically classify GPUs to ensure that they are allocated WUs which are best suited to them. There's no ETA but work is happening so let's see what happens 😃
